Avoid unsafe cast when removing style layers

boundsj opened the pull request:
This updates the code generation for style layers to add a test to check that the style object returned from mbgl can be safely cast into the expected type so ownership can be returned to the caller when removing layers.

Previously a crash occurred if a layer of type T was added before a layer of type U but both layers shared the same identifier. In that case, if the pointer to the second layer of type U
was used to remove that layer, a runtime exception would occur since mbgl returns the layer with type T (the first one added) and the unsafe cast was triggered.

Now, even though the unsafe cast is avoided, an exception will still occur if layers with the same identifier are manipulated as described above but this is done as part of a validation check for all public style layer methods at the Darwin platform level. The raised `NSException` attempts to warn developers about reusing style identifiers to avoid this issue in the first place.

Fixes https://github.com/mapbox/mapbox-gl-native/issues/7211

cc @1ec5 @jfirebaugh

1ec5 commented on platform/darwin/src/MGLStyleLayer_Private.h:
```diff
@@ -5,6 +5,25 @@
+ of `MGLStyleLayer` and its subclasses. For private methods, an assertion is more appropriate.
+ */
+#define MGLAssertStyleLayerIsValid() \
+do { \
```
Please indent this macro so it’s more readable.

1ec5 commented on platform/darwin/src/MGLStyleLayer_Private.h:
```diff
@@ -5,6 +5,25 @@
+if (!self.rawLayer) { \
+[NSException raise:@"Invalid style layer" \
+format: \
+@"-[%@ removeLayer:] has been called " \
```
The method is named `-[MGLStyle removeLayer:]`, not `-[MGL*StyleLayer removeLayer:]`.

boundsj commented:
Squashed and rebased on top of https://github.com/mapbox/mapbox-gl-native/pull/7128/commits/0c518addd08e272ee9489bfdb4fddd1a52105a08 (https://github.com/mapbox/mapbox-gl-native/pull/7128)
